strict-version-matcher-plugin/src/main/kotlin/com/google/android/gms/dependencies/Versions.kt

```diff
@@ -0,0 +1,108 @@
+         * services current use.
+         *
+         * @param[versionString] Three-part Semver string to convert.
+         */
+        fun parseString(versionString:String) : SemVerVersionInfo {
+            val version = versionString.trim();
+            val parts = version.split(".")
+            if (parts.size != 3) {
+                throw IllegalArgumentException("versionString didn't have 3 parts divided by periods.");
+            }
+            val major = Integer.valueOf(parts[0])
+            val minor = Integer.valueOf(parts[1])
+
+            var patchString = parts[2]
+            var dashIndex = patchString.indexOf("-")
+            if (dashIndex != -1) {
+                patchString = patchString.substring(0, dashIndex)
+            }
+            // TODO: Update to account for qualifiers to the version.
+            val patch = Integer.valueOf(patchString)
+            return SemVerVersionInfo(major= major, minor = minor, patch = patch)
+        }
+    }
+}
+
+data class Version(val rawString: String, val trimmedString: String) {
+    companion object {
+        fun fromString(version: String?) : Version? {
+            if (version == null) {
+                return null
+            }
+            return Version(version, version.split("-")[0])
+        }
+    }
+}
+
+data class VersionRange(val closedStart: Boolean, val closedEnd: Boolean, val rangeStart: Version, val rangeEnd:Version) {
+    fun toVersionString() : String {
+        return (if (closedStart) "[" else "(") + rangeStart.trimmedString + "," + rangeEnd.trimmedString + (if (closedEnd) "]" else ")")
+    }
+
+    fun versionInRange(compareTo : Version) : Boolean {
+        if (closedStart) {
+            if (versionCompare(rangeStart.trimmedString, compareTo.trimmedString) > 0) {
+                return false
+            }
+        } else {
+            if (versionCompare(rangeStart.trimmedString, compareTo.trimmedString) >= 0) {
+                return false
+            }
+        }
+        if (closedEnd) {
+            if (versionCompare(rangeEnd.trimmedString, compareTo.trimmedString) < 0) {
+                return false
+            }
+        } else {
+            if (versionCompare(rangeEnd.trimmedString, compareTo.trimmedString) <= 0) {
+                return false
+            }
+        }
+        return true
+    }
+
+    companion object {
+        fun versionCompare(str1 : String, str2 : String) : Int {
+            val vals1 = str1.split("\\.")
+            val vals2 = str2.split("\\.")
+            var i = 0
+            while (i < vals1.size && i < vals2.size && vals1[i] == vals2[i]) {
+                i++
+            }
+            if (i < vals1.size && i < vals2.size) {
+                val diff = Integer.compare(Integer.valueOf(vals1[i]), Integer.valueOf(vals2[i]))
+                return Integer.signum(diff)
+            }
+            return Integer.signum(vals1.size - vals2.size)
```

Comment:
Wait, is this really how Maven has defines the comparison of un-equally sized version strings?